### ISSUE TITLE
ExecutorLostFailure while training LightGBMRegressor model

### ISSUE THREAD

**Issue by Lyvz** [microsoft/SynapseML#1341]
Hi!

I'm consistently encountering issues while trying to train my model and the error messages I'm receiving aren't pointing me in the right direction. 

My initial suspicion was that the issue could be memory related, but my debugging attempts so far haven't decisively confirmed it.

Further details below:

#### SynapseML version (Maven coordinate):
com.microsoft.azure:synapseml_2.12:0.9.4

#### Cluster specs:
Databricks runtime version: Runtime: 8.4 ML (Scala 2.12, Spark 3.1.2)
Worker type: Standard_DS4_v2, autoscaling min=2, max=4
Driver type: Standard_DS4_v2

#### Training data:
22 676 857 rows
8 features
~ 330 MB (returned by org.apache.spark.util.SizeEstimator)
OR 22 676 857 rows * 8 cols * 8 bytes per col = 1451318848 ~= 1.45 GB

#### Training code:
```
lgr = LightGBMRegressor(alpha=0.9,
                        baggingFraction=0.9,
                        baggingFreq=3,
                        baggingSeed=3,
                        boostFromAverage=True,
                        boostingType='gbdt',
                        categoricalSlotNames=['feature_1'],
                        defaultListenPort=12400,
                        earlyStoppingRound=100,
                        featureFraction=0.9,
                        featuresCol='features',
                        isProvideTrainingMetric=False,
                        labelCol='target',
                        lambdaL1=0.0,
                        lambdaL2=0.0,
                        learningRate=0.1,
                        maxBin=510,
                        maxDepth=12,
                        metric='rmse',
                        minSumHessianInLeaf=0.01,
                        numBatches=0,
                        numIterations=200,
                        numLeaves=512,
                        objective='regression',
                        parallelism='data_parallel',
                        predictionCol='prediction',
                        timeout=1800.0,
                        tweedieVariancePower=1.5,
                        useBarrierExecutionMode=True,
                        verbosity=1)

pipeline = Pipeline(
    stages=[
        featurizer]
)
lr_train_data = pipeline.fit(train).transform(train)
lr_train_data = lr_train_data['target', 'features']

lr_test_data = pipeline.fit(fulltest).transform(fulltest)
lr_test_data = lr_test_data['feature_1','target', 'features']

lg_model = lgr.fit(lr_train_data)

scoredData = lg_model.transform(lr_test_data)
```
#### Stacktrace
```
org.apache.spark.SparkException: Job aborted due to stage failure: Could not recover from a failed barrier ResultStage. Most recent failure reason: Stage failed because barrier task ResultTask(12, 7) finished unsuccessfully.
ExecutorLostFailure (executor 1 exited caused by one of the running tasks) Reason: Remote RPC client disassociated. Likely due to containers exceeding thresholds, or network issues. Check driver logs for WARN messages.
	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2778)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2725)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2719)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2719)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskCompletion(DAGScheduler.scala:2476)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2983)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2927)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2915)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:1028)
	at org.apache.spark.SparkContext.runJobInternal(SparkContext.scala:2459)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2442)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2554)
	at org.apache.spark.rdd.RDD.$anonfun$reduce$1(RDD.scala:1130)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:125)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:419)
	at org.apache.spark.rdd.RDD.reduce(RDD.scala:1112)
	at com.microsoft.azure.synapse.ml.lightgbm.LightGBMBase.innerTrain(LightGBMBase.scala:484)
	at com.microsoft.azure.synapse.ml.lightgbm.LightGBMBase.innerTrain$(LightGBMBase.scala:443)
	at com.microsoft.azure.synapse.ml.lightgbm.LightGBMRegressor.innerTrain(LightGBMRegressor.scala:38)
	at com.microsoft.azure.synapse.ml.lightgbm.LightGBMBase.$anonfun$train$1(LightGBMBase.scala:63)
	at com.microsoft.azure.synapse.ml.logging.BasicLogging.logVerb(BasicLogging.scala:63)
	at com.microsoft.azure.synapse.ml.logging.BasicLogging.logVerb$(BasicLogging.scala:60)
	at com.microsoft.azure.synapse.ml.lightgbm.LightGBMRegressor.logVerb(LightGBMRegressor.scala:38)
	at com.microsoft.azure.synapse.ml.logging.BasicLogging.logTrain(BasicLogging.scala:49)
	at com.microsoft.azure.synapse.ml.logging.BasicLogging.logTrain$(BasicLogging.scala:48)
	at com.microsoft.azure.synapse.ml.lightgbm.LightGBMRegressor.logTrain(LightGBMRegressor.scala:38)
	at com.microsoft.azure.synapse.ml.lightgbm.LightGBMBase.train(LightGBMBase.scala:44)
	at com.microsoft.azure.synapse.ml.lightgbm.LightGBMBase.train$(LightGBMBase.scala:43)
	at com.microsoft.azure.synapse.ml.lightgbm.LightGBMRegressor.train(LightGBMRegressor.scala:38)
	at com.microsoft.azure.synapse.ml.lightgbm.LightGBMRegressor.train(LightGBMRegressor.scala:38)
	at org.apache.spark.ml.Predictor.fit(Predictor.scala:151)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:380)
	at py4j.Gateway.invoke(Gateway.java:295)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:251)
	at java.lang.Thread.run(Thread.java:748)
```

#### Executor log (executor 1 from SparkException stacktrace):
```
.
.
.
22/01/12 14:19:28 INFO LightGBMRegressor: LightGBM running iteration: 50 with is finished: false
22/01/12 14:19:28 INFO LightGBMRegressor: LightGBM running iteration: 50 with is finished: false
22/01/12 14:19:28 INFO LightGBMRegressor: LightGBM running iteration: 50 with is finished: false
22/01/12 14:19:28 INFO LightGBMRegressor: LightGBM running iteration: 50 with is finished: false
22/01/12 14:19:32 INFO LightGBMRegressor: LightGBM running iteration: 51 with is finished: false
22/01/12 14:19:32 INFO LightGBMRegressor: LightGBM running iteration: 51 with is finished: false
22/01/12 14:19:32 INFO LightGBMRegressor: LightGBM running iteration: 51 with is finished: false
22/01/12 14:19:32 INFO LightGBMRegressor: LightGBM running iteration: 51 with is finished: false
22/01/12 14:19:32 INFO LightGBMRegressor: LightGBM running iteration: 51 with is finished: false
22/01/12 14:19:32 INFO LightGBMRegressor: LightGBM running iteration: 51 with is finished: false
22/01/12 14:19:32 INFO LightGBMRegressor: LightGBM running iteration: 51 with is finished: false
22/01/12 14:19:33 INFO LightGBMRegressor: LightGBM running iteration: 51 with is finished: false
22/01/12 14:19:42 INFO LightGBMRegressor: LightGBM running iteration: 52 with is finished: false
22/01/12 14:19:42 INFO LightGBMRegressor: LightGBM running iteration: 52 with is finished: false
22/01/12 14:19:42 INFO LightGBMRegressor: LightGBM running iteration: 52 with is finished: false
22/01/12 14:19:42 INFO LightGBMRegressor: LightGBM running iteration: 52 with is finished: false
22/01/12 14:19:42 INFO LightGBMRegressor: LightGBM running iteration: 52 with is finished: false
22/01/12 14:19:42 INFO LightGBMRegressor: LightGBM running iteration: 52 with is finished: false
22/01/12 14:19:42 INFO LightGBMRegressor: LightGBM running iteration: 52 with is finished: false
22/01/12 14:19:42 INFO LightGBMRegressor: LightGBM running iteration: 52 with is finished: false
[end]
```

#### What I've tried so far:

- turning ```useBarrierExecutionMode``` off --> Different error (java.net.ConnectException: Connection refused)
- scaling down size of dataset --> works with slightly higher ```numIterations``` and vice versa
- using smaller ```maxDepth```, ```maxBin``` and ```numLeaves``` --> works but model accuracy is slightly compromised
- scaling up worker and driver specs in the cluster --> not helping
- using ```parallelism='voting_parallel'``` --> not helping
- using ```singleDatasetMode=True``` --> not helping
- using different training dataset partitions before ```.fit()``` --> not helping
- using ```com.microsoft.ml.spark:mmlspark_2.12:1.0.0-rc3-59-bf337941-SNAPSHOT``` --> not helping
- monitoring worker memory usage from Spark UI --> didn't go over max memory
- removing feature_1 from ```categoricalSlotNames``` --> works but model accuracy is greatly compromised

Greatly appreciate any help!

**Comment by imatiach-msft** [comment]
hi @Lyvz sorry about the error you are encountering.
"ExecutorLostFailure (executor 1 exited caused by one of the running tasks) Reason: Remote RPC client disassociated. Likely due to containers exceeding thresholds, or network issues. Check driver logs for WARN messages.
"
Indeed, this is usually due to OOM.  It can also be caused by excessive logging (there is a 20 MB log limit in databricks which can be hit easily if you log a lot) - but based on your description, since it worked for less data, it's most likely due to OOM.

"scaling up worker and driver specs in the cluster --> not helping"
I am surprised that this does not help.

"using singleDatasetMode=True --> not helping"
I would recommend to use singleDatasetMode moving forward, and that has been made the default behavior in 0.9.5.

I would recommend removing the timeout parameter (although your value does seem larger than the default 1200 which is good), otherwise that may cause the connection refused error.

"numBatches=0"
Please don't specify numBatches, unless you want to split the dataset into multiple pieces and retrain the model on each subsequent piece, I don't recommend it.

If you are still seeing issues it might be good to have a call to debug.

**Comment by imatiach-msft** [comment]
oh actually, I just noticed this:
verbosity=1
this could cause excessive logging and hit the 20 MB log limit in databricks for large data

**Comment by imatiach-msft** [comment]
"Worker type: Standard_DS4_v2, autoscaling min=2, max=4"
autoscaling is currently not supported, please disable autoscale, it can cause errors to occur if nodes are removed during training

**Comment by Lyvz** [comment]
Hi @imatiach-msft!

I ran the training code again after applying your suggested changes (except removing the timeout parameter) and the training completed without errors. That's great news! For this run I used a cluster with 4x Standard_DS4_v2 workers.

Thanks for the quick reply and helpful comments!
